src/Main.elm

```diff
@@ -1,37 +1,149 @@
-module Main exposing (Model, Msg(..), book, init, main, update, view)
+module Main exposing (Model, Msg(..), init, main, update, view)
 
 import BookData exposing (Book, allBooks)
 import Browser
-import Html exposing (Html, div, h1, img, li, text, ul)
+import Html exposing (Html, button, div, h1, img, li, table, td, text, th, thead, tr, ul)
 import Html.Attributes exposing (alt, class, src)
+import Html.Events exposing (onClick)
+import List exposing (append, filter, map, sum)
+import List.Extra exposing (setIf)
 
 
 
 ---- MODEL ----
 
 
 type alias Model =
-    { books : List Book }
+    { books : List Book, cart : Cart }
+
+
+type alias CartItem =
+    { id : BookID, quantity : Int }
+
+
+type alias Cart =
+    List CartItem
+
+
+type alias BookID =
+    Int
+
+
+addToCart : BookID -> Cart -> Cart
+addToCart bookID cart =
+    case inCart bookID cart of
+        True ->
+            let
+                incrementQuantity cartItem =
+                    if cartItem.id == bookID then
+                        { cartItem | quantity = cartItem.quantity + 1 }
+
+                    else
+                        cartItem
+            in
+            List.map (\item -> incrementQuantity item) cart
+
+        -- setIf (\cartItem -> cartItem.id == bookID) { cartItem | quantity = cartItem.quantity + 1 } cart
+        False ->
+            append cart [ { id = bookID, quantity = 1 } ]
+
+
+removeFromCart : BookID -> Cart -> Cart
+removeFromCart bookID cart =
+    List.filter (\cartItem -> cartItem.id /= bookID) cart
+
+
+increment : Cart -> BookID -> Cart
+increment cart bookID =
```

Comment:
Again, I think this and `decrement` are good candidates for `replaceIf` / `setIf`